src/SharpLearning.Backend.Cntk.Test/CntkRawMnistTest.cs

```diff
@@ -0,0 +1,575 @@
+    // Do NOT use this code as foundation for other things, quality is not great!
+
+    // Stores the per-image MNIST information we loaded from disk 
+    //
+    // We store the data in two formats, byte array (as it came in from disk), and float array
+    // where each 0..255 value has been mapped to 0.0f..1.0f
+    public struct MnistImage
+    {
+        public int Cols, Rows;
+        public byte[] Data;
+        public float[] DataFloat;
+
+        public MnistImage(int cols, int rows, byte[] data)
+        {
+            Cols = cols;
+            Rows = rows;
+            Data = data;
+            DataFloat = new float[data.Length];
+            for (int i = 0; i < data.Length; i++)
+            {
+                DataFloat[i] = Data[i] / 255f;
+            }
+        }
+    }
+
+    // Helper class used to load and work with the Mnist data set
+    public class Mnist
+    {
+        // 
+        // The loaded results
+        //
+        public MnistImage[] TrainImages, TestImages, ValidationImages;
+        public byte[] TrainLabels, TestLabels, ValidationLabels;
+        public byte[,] OneHotTrainLabels, OneHotTestLabels, OneHotValidationLabels;
+
+        public BatchReader GetTrainReader(int readFromItem) => new BatchReader(TrainImages, TrainLabels, OneHotTrainLabels, readFromItem);
+        public BatchReader GetTrainReader() => new BatchReader(TrainImages, TrainLabels, OneHotTrainLabels);
+        public BatchReader GetTestReader() => new BatchReader(TestImages, TestLabels, OneHotTestLabels);
+        public BatchReader GetValidationReader() => new BatchReader(ValidationImages, ValidationLabels, OneHotValidationLabels);
+
+        public class BatchReader
+        {
+            int start = 0;
+            MnistImage[] source;
+            byte[] labels;
+            byte[,] oneHotLabels;
+
+            internal BatchReader(MnistImage[] source, byte[] labels, byte[,] oneHotLabels, int readFromItem = 0)
+            {
+                this.source = source;
+                this.labels = labels;
+                this.oneHotLabels = oneHotLabels;
+                start = readFromItem;
+            }
+
+            public (float[,], float[,]) NextBatch(int batchSize)
+            {
+                // TODO: Remove consts and allocs...
+                var imageData = new float[batchSize, 784];
+                var labelData = new float[batchSize, 10];
+
+                int p = 0;
+                for (int item = 0; item < batchSize; item++)
+                {
+                    Buffer.BlockCopy(source[start + item].DataFloat, 0, imageData, p, 784 * sizeof(float));
+                    p += 784 * sizeof(float);
+                    for (var j = 0; j < 10; j++)
+                        labelData[item, j] = oneHotLabels[item + start, j];
+                }
+
+                start += batchSize;
+                return (imageData, labelData);
+            }
+
+            public (float[], float[]) NextBatchArray(int batchSize)
```

Comment:
Added extra method for outputting images and targets as single arrays. This matches the cntk better.